### DISCUSSION TITLE
One bar click event and y-axis label format

### DISCUSSION THREAD

**Discussion by chaeeunDev** [patrykandpatrick/vico#685] [Questions] [answered]
Hello!
I'm using this library with kotlin and jetpack compose to make bar graphs, and I've got some questions.

1. If I click on one bar, can I color only that bar differently? Can I color only the x-axis label of the clicked bar in the same context?

2. And can I get an index for the clicked bar?

3. I saw the example of value formatting for x-axis label, how do I value formatting for y-axis label? I used itemPlacer to adjust the number of y-axis label, but I want to format the y-axis label separately in the form of 'hh:mm'.

4. I want to show specific data for this month on one screen, but now there are a lot of x-axis values, so it comes out in the form of scrolling horizontally and looking at the graph?
But I want to make the stick thin and see it on the screen, I don't show all the x-axis labels at the bottom, I use the item place to show only a few labels, but if I click on one bar, I want to show that label.

5. Finally, I can put a title on the y-axis, can I change the position to the top of the y-axis?
The current state is with the title on the right side of the y-axis!

Thank you for providing a good library. If you can answer these questions, I'll refer to them and try making a graph again!

**Comment by Jacob3075** [comment]
I have some similar requirements as you, was able to get some of them done but i was not able to get the label to show up only if you click on the bar yet
bellow is for formatting the labels on the x axis and skipping some of them

  ```kotlin
   Chart(
        bottomAxis = rememberBottomAxis(
             valueFormatter = xAxisFormatter
        )
)
```

`xAxisFormatter`, this will skip every n values and format the label with Mon, Tue, Wed, etc, you can change to your use case
```kotlin
    fun getDefaultXAxisFormatter(today: LocalDate, skipCount: Int = 1): (Float, ChartValues) -> String = { value, chartValues ->
        if (value.toInt() % skipCount != 0) ""
        else
            today.minus(
                chartValues.chartEntryModel.entries.first().size - 1 - value.toLong(),
                DateTimeUnit.DAY,
            ).getDisplayNameForDay()
    }
```

but i did face 1 pretty major issue with the axis label, i was not able to disable ellipsis even tho there is a lot of space around the labels cause i'm skipping most of them, you might also face this cause your label text seems like it will take up a lot space

![image](https://github.com/patrykandpatrick/vico/assets/45714347/35cbd1bb-badd-40cb-bacd-43905d7d51d4)

if anyone is able to remove ellipsis then please update here

**Reply by Gowsky** [reply]
An empty label is still a label. You can skip x-axis labels by using `AxisItemPlacer.Horizontal`. `AxisItemPlacer.Horizontal.default` has a spacing parameter. Formatters shouldn't return empty `String`s.

**Reply by chaeeunDev** [reply]
Thank you for the detailed explanation. I think I can solve the problem!! thank you!

**Comment by muratgny** [comment]
Answer to your third question:

private val endAxisValueFormatter = CartesianValueFormatter { x, _ , _ -> "$x kW" } // Here X is the item of a list

You do not have to pass all list to the formatter. This works for me.

**Reply by chaeeunDev** [reply]
Wow, there's such a simple way. Thank you!!

**Comment by marcusfreitas** [comment]
How you guys implemented the event when tapping the Bar?

**Comment by Gowsky** [comment, marked as answer]
Hello, @chaeeunDev.

> If I click on one bar, can I color only that bar differently?

If you're using Vico 2, yes. Please see #769.

> Can I color only the x-axis label of the clicked bar in the same context?

You can do this using `CartesianValueFormatter` (`AxisValueFormatter` in Vico 1) and `SpannableString`. Selection would work analogously — you'd obtain the _x_ value of the tapped column via `CartesianMarkerVisibilityListener` (`MarkerVisibilityChangeListener` in Vico 1).

> And can I get an index for the clicked bar?

Yes, please see #769 and #845. As mentioned in the latter, a dedicated, marker-independent API for selection will be introduced in the future. (To reiterate, if you're using Vico 1, `CartesianMarkerVisibilityListener` is called `MarkerVisibilityChangeListener` there.)

> I saw the example of value formatting for x-axis label, how do I value formatting for y-axis label? I used itemPlacer to adjust the number of y-axis label, but I want to format the y-axis label separately in the form of 'hh:mm'.

The approach is the same as with `HorizontalAxis`. VerticalAxis also has a `valueFormatter` property.

> I want to show specific data for this month on one screen, but now there are a lot of x-axis values, so it comes out in the form of scrolling horizontally and looking at the graph?
> But I want to make the stick thin and see it on the screen

You can do this by turning off scrolling. The zoom factor will be adjusted automatically.

> but if I click on one bar, I want to show that label.

You could do this via `ItemPlacer`, but it may be easier to achieve with a marker.

> Finally, I can put a title on the y-axis, can I change the position to the top of the y-axis?

`VerticalAxis` doesn't have such a feature at the moment. However, you can implement this behavior yourself:

- In Vico 2, you can achieve this by subclassing `VerticalAxis`.
- In Vico 1, you'd have to create a custom `AxisRenderer` based on `VerticalAxis`.

**Note that there's a recomposition issue related to strong skipping mode in version `2.0.0-beta.1`** (#900), and it may get in the way. It will be fixed in the next update.

**Reply by chaeeunDev** [reply]
Thank you so much for answering all the questions! I think I can make the graph I wanted by referring to those answers. It's a really good library, thank you!
